util/tracecpu/collector.go

```diff
@@ -0,0 +1,333 @@
+	stream, err := client.CollectCPUTime(ctx)
+	if err != nil {
+		cancel()
+		return nil, nil, nil, err
+	}
+	return conn, stream, cancel, nil
+}
+
+// NewTopSQLCollector creates a new TopSQL struct
+//
+// planBinaryDecoder is a decoding function which will be called asynchronously to decode the plan binary to string
+// maxSQLNum is the maximum SQL and plan number, which will restrict the memory usage of the internal LFU cache
+func NewTopSQLCollector(config *TopSQLCollectorConfig) *TopSQLCollector {
+	normalizedSQLMap := make(map[string]string)
+	normalizedPlanMap := make(map[string]string)
+	planRegisterChan := make(chan *planRegisterJob, 10)
+	topSQLMap := make(map[string]*TopSQLDataPoints)
+
+	ts := &TopSQLCollector{
+		topSQLMap:         topSQLMap,
+		maxSQLNum:         config.MaxSQLNum,
+		normalizedSQLMap:  normalizedSQLMap,
+		normalizedPlanMap: normalizedPlanMap,
+		planRegisterChan:  planRegisterChan,
+		agentGRPCAddress:  config.AgentGRPCAddress,
+		instanceID:        config.InstanceID,
+		quit:              make(chan struct{}),
+	}
+
+	go ts.registerNormalizedPlanWorker(config.PlanBinaryDecoder)
+
+	go ts.sendToAgentWorker(config.SendToAgentInterval)
+
+	return ts
+}
+
+// Collect uses a hashmap to store records in every minute, and evict every minute.
+// This function can be run in parallel with snapshot, so we should protect the map operations with a mutex.
+func (ts *TopSQLCollector) Collect(timestamp uint64, records []TopSQLRecord) {
+	for _, record := range records {
+		encodedKey := encodeCacheKey(record.SQLDigest, record.PlanDigest)
+		entry, exist := ts.topSQLMap[encodedKey]
+		if !exist {
+			entry = &TopSQLDataPoints{
+				CPUTimeMsList: []uint32{record.CPUTimeMs},
+				TimestampList: []uint64{timestamp},
+			}
+			ts.topSQLMap[encodedKey] = entry
+		} else {
+			entry.CPUTimeMsList = append(entry.CPUTimeMsList, record.CPUTimeMs)
+			entry.TimestampList = append(entry.TimestampList, timestamp)
+		}
+		entry.CPUTimeMsTotal += uint64(record.CPUTimeMs)
+	}
+
+	if len(ts.topSQLMap) <= ts.maxSQLNum {
+		return
+	}
+
+	// find the max CPUTimeMsTotal that should be evicted
+	digestCPUTimeList := make([]digestAndCPUTime, len(ts.topSQLMap))
+	{
+		i := 0
+		for key, value := range ts.topSQLMap {
+			data := digestAndCPUTime{
+				Key:            key,
+				CPUTimeMsTotal: value.CPUTimeMsTotal,
+			}
+			digestCPUTimeList[i] = data
+			i++
+		}
+	}
+	// QuickSelect will only return error when the second parameter is out of range
+	if err := quickselect.QuickSelect(digestAndCPUTimeSlice(digestCPUTimeList), ts.maxSQLNum); err != nil {
+		//	skip eviction
+		return
+	}
+	shouldEvictList := digestCPUTimeList[ts.maxSQLNum:]
+	for _, evict := range shouldEvictList {
+		delete(ts.topSQLMap, evict.Key)
+		sqlDigest, planDigest := decodeCacheKey(evict.Key)
+		ts.mu.Lock()
+		delete(ts.normalizedSQLMap, sqlDigest)
+		delete(ts.normalizedPlanMap, planDigest)
+		ts.mu.Unlock()
+	}
```

Comment:
Maybe we can refer to TiKV implementation that the eviction happens periodically, e.g. every minute?

Reply:
Then the memory usage may be more jittering, and the time seems not enough for the coding and testing.
Is this necessary in this first version?